File: Common/ListShifter.vb
Public Class ListShifter(Of T)
    Implements IList(Of T)

    Private ReadOnly InnerList As IList(Of T),
                     Shift As Integer

    Public Sub New(ByVal List As IList(Of T), ByVal Shift As Integer)
        Me.InnerList = List
        Me.Shift = Shift
    End Sub

    Public ReadOnly Property Count As Integer Implements ICollection(Of T).Count
        Get
            Return Me.InnerList.Count
        End Get
    End Property

    Public ReadOnly Property IsReadOnly As Boolean Implements ICollection(Of T).IsReadOnly
        Get
            Return Me.InnerList.IsReadOnly
        End Get
    End Property

    Default Public Property Item(index As Integer) As T Implements IList(Of T).Item
        Get
            Return Me.InnerList.Item(index + Me.Shift)
        End Get
        Set(value As T)
            Me.InnerList.Item(index + Me.Shift) = value
        End Set
    End Property

    Public Sub Add(item As T) Implements ICollection(Of T).Add
        Me.InnerList.Add(item)
    End Sub

    Public Sub Clear() Implements ICollection(Of T).Clear
        Me.InnerList.Clear()
    End Sub

    Public Sub CopyTo(array() As T, arrayIndex As Integer) Implements ICollection(Of T).CopyTo
        Me.InnerList.CopyTo(array, arrayIndex)
    End Sub

    Public Sub Insert(index As Integer, item As T) Implements IList(Of T).Insert
        Me.InnerList.Insert(index + Me.Shift, item)
    End Sub

    Public Sub RemoveAt(index As Integer) Implements IList(Of T).RemoveAt
        Me.InnerList.RemoveAt(index + Me.Shift)
    End Sub

    Public Function Contains(item As T) As Boolean Implements ICollection(Of T).Contains
        Return Me.InnerList.Contains(item)
    End Function

    Public Function GetEnumerator() As IEnumerator(Of T) Implements IEnumerable(Of T).GetEnumerator
        Return Me.InnerList.GetEnumerator()
    End Function

    Public Function IndexOf(item As T) As Integer Implements IList(Of T).IndexOf
        Return Me.InnerList.IndexOf(item) - Me.Shift
    End Function

    Public Function Remove(item As T) As Boolean Implements ICollection(Of T).Remove
        Return Me.InnerList.Remove(item)
    End Function

    Private Function IEnumerable_GetEnumerator() As IEnumerator Implements IEnumerable.GetEnumerator
        Return Me.GetEnumerator()
    End Function
End Class

File: Collections/ListShifter.cs
using System.Collections;
using System.Collections.Generic;

namespace Ks.Common
{
    public class ListShifter<T> : IList<T>
    {
        private readonly IList<T> InnerList;
        private readonly int Shift;

        public ListShifter(IList<T> List, int Shift)
        {
            this.InnerList = List;
            this.Shift = Shift;
        }

        public int Count
        {
            get
            {
                return this.InnerList.Count;
            }
        }

        public bool IsReadOnly
        {
            get
            {
                return this.InnerList.IsReadOnly;
            }
        }

        public T this[int index]
        {
            get
            {
                return this.InnerList[index + this.Shift];
            }
            set
            {
                this.InnerList[index + this.Shift] = value;
            }
        }

        public void Add(T item)
        {
            this.InnerList.Add(item);
        }

        public void Clear()
        {
            this.InnerList.Clear();
        }

        public void CopyTo(T[] array, int arrayIndex)
        {
            this.InnerList.CopyTo(array, arrayIndex);
        }

        public void Insert(int index, T item)
        {
            this.InnerList.Insert(index + this.Shift, item);
        }

        public void RemoveAt(int index)
        {
            this.InnerList.RemoveAt(index + this.Shift);
        }

        public bool Contains(T item)
        {
            return this.InnerList.Contains(item);
        }

        public IEnumerator<T> GetEnumerator()
        {
            return this.InnerList.GetEnumerator();
        }

        public int IndexOf(T item)
        {
            return this.InnerList.IndexOf(item) - this.Shift;
        }

        public bool Remove(T item)
        {
            return this.InnerList.Remove(item);
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return this.GetEnumerator();
        }
    }
}
